File: codeforces/611/D/D.hs
module Main where

import Data.Function (fix)

p::Integer
p = 10^9 + 7

slice from to xs = take (to - from) (drop from xs)

ways::Integer->Int
ways x = numWays (show x) 0 1

numWays::String->Int->Integer->Int
-- number of ways to split s[a:] where the first thing is at least minimumStarting
numWays s a minimumStarting
  | length s == a = 1
  | s !! a == '0' = 0
  | otherwise = sum (map f [(a+1)..(length s)])
  where
    f i
      | firstPart >= minimumStarting = numWays s i (minimumStarting + 1)
      | otherwise = 0
      where firstPart = read (slice a i s)


main = do
  print (slice 1 2 "abcdef")
  _ <- getLine
  s <- readLn
  print (ways s)

File: codeforces/611/D/D.py
#!/usr/bin/env python3
import sys, itertools

P = int(10e9) + 7

input()
S = input()

n = len(S)

memo = {}

def numWays(a, minimumStarting):
    # number of ways to split S[a:] where the first thing is at least minimumStarting

    if (a == n):
        return 1
    if (S[a] == "0"):
        return 0

    key = (a, minimumStarting)
    if key in memo:
        return memo[key]

    ans = 0
    for i in range(a+1, n+1):
        firstPart = int(S[a:i], 10)
        if firstPart >= minimumStarting:
            ans = (ans + numWays(i, firstPart+1)) % P

    memo[key] = ans
    return ans

print(numWays(0, 1))
